src/python/pants/util/objects.py

```diff
@@ -333,22 +337,27 @@ def iterate_enum_variants(cls):
       """
       return cls._singletons.values()
 
+    @classmethod
+    def register_option(cls, register, *args, **kwargs):
```

Comment:
It seems like

`register(type=MyEnum, ...)` 

would be more consistent, and thus clearer, than 

`MyEnum.register_option(...I have to pass in register, but no type here, weird...)`

Also this creates a "moral circular dependency" between this class and the options package. There's no reason this code should "know" anything about options, even if technically it doesn't introduce a build graph dependency.

Reply:
I agree, in #7233 I was thinking we would instead add the ability to extract `choices` from a `type` automagically, and thought this would be less intrusive (but I see the point that this code really shouldn't know anything about options). A simple alternative is to just require that enum options explicitly add `choices=MyEnumType.iterate_enum_variants()` during registration, but it might be nice to specify that if the value of a `type` argument to `register()` has some attribute (e.g. `choices`), that the choices could be extracted automatically.

I think the right way to go in this PR (echoing @Eric-Arellano above as well) might be to use the normal `register()` method, and just require explicitly passing `choices=MyEnumType.iterate_enum_variants()` for now, and focus this PR on figuring out the `idempotent_type_check` issue.

Reply:
Why not have the options registrar do the right thing (i.e., set `choices`) when it notices that the type is an enum?

Reply:
I have modified the `check()` method defined in `Parser#_compute_value()` to use the result of `.iterate_enum_values()` if there is a `type` arg that has that attribute!